### ISSUE TITLE
CU-Migrate Notification Slice to React Query

### ISSUE THREAD

**Issue by Sparowhawk** [department-of-veterans-affairs/va-mobile-app#8253]
Functionality should be the same

**Comment by ala-yna** [comment]
Blocked - waiting on HSP

**Comment by TKDickson** [comment]
Per notes in prep work ticket #9040, I've come up with [a planned test run for the notifications RQ migration](https://dsvavsp.testrail.io/index.php?/runs/view/5728&group_by=cases:section_id&group_order=asc) that is ready for test plan review.

**Comment by TKDickson** [comment]
Check if #7954 is also fixed by this migration (if not; nbd - would just keep ticket open for later)

**Comment by TKDickson** [comment]
DSVA support thread is resolved and I've confirmed I can successfully register for notifications, as well as fetch preferences, from upstream servers. This is unblocked and I can start working on it tomorrow.

**Comment by TKDickson** [comment]
Am not receiving PN in staging; have opened up another thread in DSVA slack to get assistance finding where the push notifications are going astray.

**Comment by TKDickson** [comment]
OK, push notifications appear to be back up and running in staging (I've at least consistently received several push notifications on an iOS device over the last 15 minutes, although not one with this branch's build), so this trio of tickets is once more unblocked. I should be able to jump into testing for them tomorrow.

**Comment by TKDickson** [comment]
......I can't get push notifications in staging for Android. Have reached out to FE Eng associated with this ticket for ideas. Here's what I've got:

- We have been able to test Android push notifications in staging in the past ([ex of a ticket that talks about testing them](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/3608)) and both Rachael and I have memories of it as well
- I am receiving push notifications in staging for iOS (after lots of troubleshooting, see above for previous blocking comments)
- I have been trying on my Pixel (Android 14) and OnePlus (Android 11), so it doesn't seem to be linked to the Android 13+ PN changes. 
- I have tried the RQ PN branch, today's dev build, the 2.33 and 2.34 RC branches, and the daily develop builds from 8/12 and 7/30, so it doesn't seem to be linked to the RQ changes, or changes within (relatively) recent timeframes.
- afaik, Android devices are still getting push notifications in production because the vama_notification_click event is still firing for Android users

Remaining cases on this trio of tickets (this, #7954, and #7926 which are all on the same branch) are "send a push notification and check XYZ thing", so they're back to blocked.

**Comment by TKDickson** [comment]
Oh, and some cases that were going to be in testing scope, will be out of scope specifically for Android as well because of #9521. That's more of a "disappointing but oh well" whereas the above is a true blocker.

**Comment by TKDickson** [comment]
As of me leaving for vacation for a couple days, this is still blocked awaiting VEText team help (there's been progress & help already, and Android is still not working in staging)

**Comment by TKDickson** [comment]
Alright, it took nearly 3 weeks of troubleshooting across several teams; but push notifications are working for both iOS and Android in staging once again. Removed the blocked-QA label from this, and the 2 related, tickets.

**Comment by TKDickson** [comment]
Issues found while testing:
- Push notifications right now - for lack of a better term - are too 'sticky'. I sent a messaging push, and used it to deeplink into the app on a few different devices. Then on one of those devices I was logged out (not manually, but came back to the app and had been kicked out), and when I logged back in, it deeplinked me to that same message. On another device, I logged out manually and logged back in as a DIFFERENT user, and it also tried to deeplink me in to that same message - but displayed an error page, since that user didn't have access to the message (from the original push). Have confirmed that this is also happening on the 2.36 RC branch, will be writing up a bug ticket and drop the link in here.
- Android only - push notifications received with the app in the foreground, completely fail to deeplink - they instead bring you to the home screen. (have confirmed this does not happen on the 2.36 RC branch, which successfully deeplinks to the intended message; also doesn't happen in RQ branch with iOS). Also, when I attempted to log out of the app, then log back in again (thinking maybe I could get deeplinking to work somewhere in there), my fingerprint login failed and I was dropped back to the UN/PW signin, so I should test around that area after getting the fix for this.

**Comment by TKDickson** [comment]
No change. All other methods of receiving a push notification still deeplink you correctly (iOS and Android), and when the app is in the foreground for Android, push notifcations 1) bring you to the home page and 2) don't 'disappear' after tapping them. I didn't mention that behavior in my original write-up, but maybe hearing about it will spark something about how to fix it.

**Comment by TKDickson** [comment]
@theodur unfortunately the deep linking is still broken for Android when the app is in the foreground, see prior two comments for more details. Also our staging user is semi-broken ([see thread](https://adhoc.slack.com/archives/C067S0FRP0C/p1726778828300589)), but you can still test push notifications with messaging.
Sending back for further investigation.

**Comment by TKDickson** [comment]
@theodur @Sparowhawk per DM, adding more documentation about where I'm seeing the Android app-in-foreground issue:
I have now tried receiving push notifications with the app in the foreground on the 2.36 RC build (from original writeup), the 9.12 develop build (pre-RN), the 9.16 develop build (post-RN), today's develop build, and the latest build from the RQ notification migration branch. The *only* place where I see this issue, is the RQ migration branch. 
Here's [a side-by-side video of the migration branch and today's develop build, where only this branch has the issue](https://drive.google.com/file/d/1O6fc6L439lCnsGMqlHuLbDL6eOda3jLi/view?usp=sharing).

**Comment by TKDickson** [comment]
The same scenario (android, app in foreground) still broken with the latest changes from yesterday

**Comment by TKDickson** [comment]
Further troubleshooting/testing currently blocked by MHV staging environment downtime due to Hurricane Helene

**Comment by TKDickson** [comment]
The troubleshooting build didn't work at all in iOS (the popup didn't appear after tapping the push, just threw an error on the messaging screen).

For Android, I was able to get screenshots for the push-received-while-app-in-background and push-received-while-app-in-foreground workflows. The push-received-while-app-closed workflow didn't work to get the popup either (deeplinked correctly when I tapped on the push and logged in, but popup didn't appear).

Here's the screenshot for 'in background' (correctly deeplinked):
<img width="361" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNmdRQnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--02eff9a942aa0087b4e60192efe3fcb2469beb51/image.png" alt="image.png" />

Here's the screenshot for 'in foreground' (not correctly deeplinking):
<img width="740" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNmtRQnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--d4fe337161cf43e7ab3e94f3ad49459a91b31d7c/image.png" alt="image.png" />

**Comment by TKDickson** [comment]
To make it super clear what I tested pre-fix and post-fix (which is *fingers crossed so hard* definitely going to work), I've closed my original test run and [spun up a second one for this round of testing.](https://dsvavsp.testrail.io/index.php?/runs/view/5989&group_by=cases:section_id&group_order=asc)

**Comment by TKDickson** [comment]
@Sparowhawk this is still failing for the same case (and passing all other cases), but failing in a slightly new way now.

Similar problem to before, but different.

iOS is passing all attempted (foreground and background). Android is passing some (background and forceclosed), and failing the same one as before (foreground) -- but when you tap on a push notification from the VA app, while you're logged into the VA app with it in the foreground/currently open, it is now requiring re-authentication via fingerprint, then bringing to home screen. Previously it was just bringing you to the home screen without requiring re-authentication. (I haven't tried this with a user with biometrics disabled, but I assume it would require me to log in with UN/PW in that scenario, if tapping a push with app in foreground)

**Comment by Sparowhawk** [comment]
@TKDickson I updated it again and hopefully it should work this time. It turns out that the code I removed was actually needed. As it was in the auth slice and that is why I didn't see it in the direct 1 to 1 comparison of the Notification Manager and App.tsx files

So as a result I have updated the UseEffect in App.tsx to act differently to more accurately take into account foreground notifications being pressed or not

**Comment by Sparowhawk** [comment]
@TKDickson I have added a patch to the library we use. Hopefully this fixes it. Otherwise we will likely have to think on migrating to a more maintained library.

**Comment by TKDickson** [comment]
The latest (of many builds) today finally successfully deeplinked from a push in the foreground on an Android device! That's big, that's the first time it's worked on this branch.

I still have more testing to do, and not everything is perfect yet.
* On the Android device where I received the push notification in the foreground, after I tapped on the push (and was brought successfully to the message via deeplinking), the push notification lingered in my list of notifications (see screenshot below). It was the only device where the push notification had that behavior (background Android, and foreground and background iOS, the push dropped off the list after tapped). Tapping the already-visited push notification, in the list of notifications, did nothing (did not change the page of the app I was on/did not deeplink back to that same message).
<img width="1354" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeG9nQnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--277bfe0e0b5cbc7f239e8bc8df033d131c2023b1/image.png" alt="image.png" />

I will continue to work through all of the cases in the run linked [in this comment](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/8253#issuecomment-2388927042); and will add anything else I find from that, as further comments in this ticket. Will be sending the ticket back/commenting on the PR, once I've done all the testing I want to do on this build.

**Comment by TKDickson** [comment]
OK, finished the planned testing on this branch. I'm still seeing known issue #9612 (where deeplinking persists through logout and logging back in as a new user, very weird). I also am seeing this issue:
* On Android only (iOS is fine), a user who logs out (and has not logged back in again) is still receiving push notifications on the device where they logged out. This is a reintroduction of bug #7229, although I can't tell you if it's the same root cause or not. And, once again, I have no idea why iOS is fine and Android is broken. Sadly this issue is also present in develop, so I will be writing a ticket up for that.

**Comment by TKDickson** [comment]
I have also confirmed that the "push when app in foreground doesn't go away" is *also* present in develop, so I'll spin a ticket up for that as well.

**Comment by TKDickson** [comment]
I wanted to test that this merged in well, since it was such a large change, but the MHV staging outage due to hurricane will prevent that from happening before the end of sprint. Closing all tickets associated with this PR
